### PR TITLE
ci-k8sio-audit periodic job

### PR DESCRIPTION
**Pull request by hh**
Fixes https://github.com/kubernetes/k8s.io/issues/244

Developing using https://github.com/kubernetes/test-infra/blob/master/prow/build_test_update.md#using-pj-on-kindsh

### CODE REVIEW

**Comment by hh** [comment]
I’m trying to use ./config/pj-on-kind.sh to verify a job that needs to use gcloud. It will eventually run in run in the k8s-infra-prow-build-trusted cluster and
use the k8s-infra-gcp-auditor GCP service account via workload identity. I’m trying to figure how how to pass my own GCP , gcloud auth login generated creds into the local prow job so it has access to run  gcloud w/ audit permissions.... will cross post this to #sig-testing in slack.

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 1]
```diff
@@ -22,6 +22,38 @@ periodics:
...
       - "cd groups && make run -- --confirm"
+- name: ci-k8sio-audit
+  interval: 6h
+  cluster: k8s-infra-prow-build-trusted
+  decorate: true
+  max_concurrency: 1
+  annotations:
+    testgrid-dashboards: wg-k8s-infra-k8sio
+    testgrid-alert-email: k8s-infra-alerts@kubernetes.io
+    testgrid-num-failures-to-alert: '1'
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:33]: I would change these while you're iterating on this job

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 2]
```diff
@@ -22,6 +22,38 @@ periodics:
...
+      - bash
+      args:
+      - -c
+      - |
+        gcloud config list
+        cd audit
+        ./audit-gcp.sh
+        git diff
+        git add --all
+        git commit -m "audit: update as of $(date +%Y-%m-%d)"
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:56]: As a start, you could have this job fail if there are changes to commit, and pass if there are none

**Review by spiffxp — CHANGES_REQUESTED** [review]
Some suggestions

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 3]
```diff
@@ -22,6 +22,80 @@ periodics:
...
+    containers:
+    - name: groups
+      image: gcr.io/k8s-testimages/gcb-docker-gcloud:v20201130-750d12f
+      command:
+      - bash
+      args:
+      - -c
+      - |
+        echo "Some initial debuging to figure out what it looks like here" >&2
+        pwd
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:52]: >periodic jobs will have the working directory set to the root of the repo specified by the first ref in extra_refs (if specified)

ref: https://github.com/kubernetes/test-infra/blob/master/prow/pod-utilities.md

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 4]
```diff
@@ -22,6 +22,80 @@ periodics:
...
+        echo "Ensure gcloud creds are working" >&2
+        gcloud config list
+        echo -n "Calculate github user from token: " >&2
+        GH_TOKEN=$(cat /etc/github-token/oauth)
+        GH_USER=$(curl -H "Authorization: token $GH_TOKEN" "https://api.github.com/user" 2>/dev/null | sed -n "s/\s\+\"login\": \"\(.*\)\",/\1/p")
+        FORK_GH_BRANCH=autoaudit-${PROW_INSTANCE_NAME:-prow}
+        echo "$GH_USER" >&2
+        echo "Running Audit Script to dump GCP configuration to yaml" >&2
+        cd ../audit
+        ./audit-gcp.sh
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:63]:
```suggestion
        push ./audit
        ./audit-gcp.sh
        popd
```

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 5]
```diff
@@ -22,6 +22,80 @@ periodics:
...
+        echo "Generate pr-creator binary from k/test-infra/robots" >&2
+        pushd ../test-infra
+        go build -o ../k8s.io/pr-creator robots/pr-creator/main.go
+        popd
+
+        echo "Prepare commit for possible PR"
+        # Values pulled from config/prow/autobump-config.yaml
+        # and logic from prow/cmd/autobump.sh
+        git config user.name "Kubernetes Prow Robot"
+        git config user.email "k8s.ci.robot@gmail.com"
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:74]: This should correspond to the GH_USER being used, and I would advise we not use k8s.ci.robot

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 6]
```diff
@@ -22,6 +22,80 @@ periodics:
...
+      - |
+        echo "Some initial debuging to figure out what it looks like here" >&2
+        pwd
+        ls
+        echo "Ensure gcloud creds are working" >&2
+        gcloud config list
+        echo -n "Calculate github user from token: " >&2
+        GH_TOKEN=$(cat /etc/github-token/oauth)
+        GH_USER=$(curl -H "Authorization: token $GH_TOKEN" "https://api.github.com/user" 2>/dev/null | sed -n "s/\s\+\"login\": \"\(.*\)\",/\1/p")
+        FORK_GH_BRANCH=autoaudit-${PROW_INSTANCE_NAME:-prow}
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:59]: not blocking but what's the purpose of the PROW_INSTANCE_NAME substitution?

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 7]
```diff
@@ -22,6 +22,80 @@ periodics:
...
+
+        echo "Prepare commit for possible PR"
+        # Values pulled from config/prow/autobump-config.yaml
+        # and logic from prow/cmd/autobump.sh
+        git config user.name "Kubernetes Prow Robot"
+        git config user.email "k8s.ci.robot@gmail.com"
+        git add --all -u
+        git commit -m "audit: update as of $(date +%Y-%m-%d)"
+        echo -e "Pushing commit to github.com/${GH_USER}/${FORK_GH_REPO}:..." >&2
+        git push -f "https://${GH_USER}:@github.com/${GH_USER}/${FORK_GH_REPO}" "HEAD:${FORK_GH_BRANCH}}" 2>/dev/null
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:78]: I think this is going to require a token?
```suggestion
        git push -f "https://${GH_USER}:${GH_TOKEN}@github.com/${GH_USER}/${FORK_GH_REPO}" "HEAD:${FORK_GH_BRANCH}}" 2>/dev/null
```

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 8]
```diff
@@ -22,6 +22,80 @@ periodics:
...
+
+        echo "Creating PR to merge ${GITHUB_USER}:${FORK_GH_BRANCH} into master..." >&2
+        title="audit: update as of $(date +%Y-%m-%d)"
+        body="Audit Updates\n/wg k8s-infra"
+        /pr-creator \
+          --github-token-path=/etc/github-token/oauth \
+          --org=kubernetes --repo=k8s.io --branch=master \
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:85]: We just renamed this
```suggestion
          --org=kubernetes --repo=k8s.io --branch=main \
```

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 9]
```diff
@@ -22,6 +22,80 @@ periodics:
...
+          --body="${body}" \
+          --confirm
+      volumeMounts:
+      - name: github
+        mountPath: /etc/github-token
+        readOnly: true
+    volumes:
+    - name: github
+      secret:
+        secretName: oauth-token
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:98]: I need to double check if we have this.  If we don't, I suggest we not use k8s-ci-robot, and for the short-term use a user under your control (using fejta-bot as the model, which is not an org member)

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 10]
```diff
@@ -22,6 +22,80 @@ periodics:
...
+    testgrid-alert-email: hh@ii.coop
+    testgrid-num-failures-to-alert: '100'
+  extra_refs:
+  - org: kubernetes
+    repo: k8s.io
+    base_ref: master
+  - org: kubernetes
+    repo: test-infra
+    base_ref: master
+  spec:
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:41]: We should consider a github team to use for this instead, but for now we want to make sure you can re-run this for troubleshooting
```suggestion
  rerun_auth_config:
    github_users:
      - hh
  spec:
```

**Review by spiffxp — CHANGES_REQUESTED** [review]
Some suggestions

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml (hunk 3 above)
Comment (hh) [reply config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:55]: Thanks, that's useful!

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml (hunk 6 above)
Comment (hh) [reply config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:62]: It's a bit of copy-pasta from here:
https://github.com/kubernetes/test-infra/blob/ff516b7b3a904bc4500c7f859d51681161680f5b/prow/cmd/autobump/autobump.sh#L29-L31

Possibly not necessary as I don't expect there to be multiple FORK_GH_BRANCH, but I suspect this repo does host multiple prow instances.

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml (hunk 9 above)
Comment (hh) [reply config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:102]: Hmm... I'm not sure how to load secrets like this into the production prow instance.
I generated an `oauth-token` for @cncf-ci and can send that in a secure way as soon as I learn how. :)

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml (hunk 2 above)
Comment (hh) [reply config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:77]: I may swing back around to that. Trying to get a full run to ensure all the tooling works.

**Comment by hh** [comment]
Created a google secret in the format of a k8s secret in the test-pods namespace:

```
kubectl create secret  --dry-run=client -n test-pods generic cncf-ci-github-token --from-literal=token=${GH_TOKEN} -o yaml | gcloud secrets versions add --project=k8s-infra-prow-build-trusted cncf-ci-github-token --data-file=-
```

```
Created version [1] of the secret [cncf-ci-github-token].
```

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml (hunk 10 above)
Comment (hh) [reply config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:44]: How would I go about triggering a rerun for a periodic job?

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml (hunk 9 above)
Comment (spiffxp) [reply config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:99]:
```suggestion
        secretName: cncf-ci-github-token
```
```bash
gcloud secrets versions access \
  --project=k8s-infra-prow-build-trusted \
  --secret=cncf-ci-github-token \
  latest | \
    kubectl --context=gke_k8s-infra-prow-build-trusted_us-central1_prow-build-trusted \
      apply -f -
```
```
secret/cncf-ci-github-token created
```

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 11]
```diff
@@ -22,6 +22,81 @@ periodics:
...
+      image: gcr.io/k8s-testimages/gcb-docker-gcloud:v20201130-750d12f
+      command:
+      - bash
+      args:
+      - -c
+      - |
+        echo "Ensure gcloud creds are working" >&2
+        gcloud config list
+        echo -n "Calculate github user from token: " >&2
+        GH_TOKEN=$(cat /etc/github-token/oauth)
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:57]:
```suggestion
        GH_TOKEN=$(cat /etc/github-token/token)
```

**Review by spiffxp — CHANGES_REQUESTED** [review]
Added the secret @hh transferred to me via secret manager, job needs some config updates to match name of/in the secret

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml (hunk 11 above)
Comment (hh) [reply config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml]: Fixed in https://github.com/kubernetes/test-infra/pull/20742/commits/7663f609e19d2a40ce2bb38db348a28fc29f909c

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm
/wg k8s-infra
/sig testing

File: config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml [hunk 12]
```diff
@@ -22,6 +22,80 @@ periodics:
...
+  decorate: true
+  max_concurrency: 1
+  annotations:
+    testgrid-dashboards: wg-k8s-infra-k8sio
+    testgrid-alert-email: hh@ii.coop
+    testgrid-num-failures-to-alert: '100'
+  extra_refs:
+  - org: kubernetes
+    repo: k8s.io
+    base_ref: master
```
Comment (spiffxp) [inline config/jobs/kubernetes/wg-k8s-infra/trusted/wg-k8s-infra-trusted.yaml:37]: https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-k8sio-audit/1359969296511406080

d'oh, this should have been `main` not `master`, my bad for missing this on review

**Pull request merged**